eva/expression/function_expression.py

```diff
@@ -157,6 +173,65 @@ def _gpu_enabled_function(self):
                     self._function_instance = self._function_instance.to_device(device)
         return self._function_instance
 
+    def _apply_function_expression(self, func: Callable, batch: Batch, **kwargs):
+        """
+        If cache is not enabled, call the func on the batch and return.
+        If cache is enabled:
+        (1) iterate over the input batch rows and check if we have the value in the
+        cache;
+        (2) for all cache miss rows, call the func;
+        (3) iterate over each cache miss row and store the results in the cache;
+        (4) stitch back the partial cache results with the new func calls.
+        """
+        func_args = Batch.merge_column_wise(
+            [child.evaluate(batch, **kwargs) for child in self.children]
+        )
+
+        if not self._cache:
+            return func_args.apply_function_expression(func)
+
+        cols = [obj.name for obj in self.output_objs]
+
+        # 1. check cache
+        # We are required to iterate over the batch row by row and check the cache.
+        # This can hurt performance, as we have to stitch together columns to generate
+        # row tuples. Is there an alternative approach we can take?
+
+        results = np.full([len(batch), len(cols)], None)
+        keys = batch
+        if self._cache.key:
```

Comment:
When I first read code, I was thinking this is the key of the cache store. Maybe add a comment. This is to optimize the key for the function expression (data -> id).